### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by andrewmadden**

### CODE REVIEW

File: mod_form.php [hunk 1]
```diff
@@ -65,6 +65,29 @@ protected function definition() {
...
+     * @param array $files array of uploaded files "element_name"=>tmp_file_path
+     * @return array of "element_name"=>"error_description" if there are errors,
+     *         or an empty array if everything is OK (true allowed for backwards compatibility too).
+     */
+    public function validation($data, $files) {
+        $errors = parent::validation($data, $files);
+        if (isset($data['apptypeid'])) {
+            if (empty($data['apptypeid'])) {
+                $errors['apptypeid'] = get_string('error:apptypenotselected', 'applaunch');
+            } else if (\mod_applaunch\app_type::get_record(['id' => $data['apptypeid'], 'enabled' => 1]) === false) {
```
Comment (Peterburnett) [inline mod_form.php:84]: More a personal style question, any reason its not `else if (!apptype::get_record...)

Comment (andrewmadden) [reply mod_form.php:84]: Yeah, if it returned a bool only then I would maybe do this. I prefer this pattern with PHP due to it's type looseness. I only want this condition to be true if it actually returns false as opposed to falsey values.

File: tests/phpunit/applaunch_test.php [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+ */
+
+use mod_applaunch\app_type;
+use mod_applaunch\applaunch;
+
+defined('MOODLE_INTERNAL') || die();
+
+global $CFG;
+
+require_once($CFG->dirroot . '/mod/applaunch/lib.php');
```
Comment (Peterburnett) [inline tests/phpunit/applaunch_test.php]: Any reason this isn't autoloaded? not sure if this is a mod specific thing...

File: mod_form.php (hunk 1 above)
Comment (andrewmadden) [reply mod_form.php:84]: 99% of the time !function() would probably have the same behaviour, but it's like the pattern `strpos !== false`. I would rather be more specific.

File: tests/phpunit/applaunch_test.php (hunk 2 above)
Comment (andrewmadden) [reply tests/phpunit/applaunch_test.php]: It's not a class, as it's the main lib file. I don't usually touch them so wasn't sure if they were autoloaded or not?

Comment (Peterburnett) [reply tests/phpunit/applaunch_test.php]: lib.php specifically has some magic afaik, along with version.php. Try it and see?

Comment (andrewmadden) [reply tests/phpunit/applaunch_test.php]: You're right. Good to know. Removed require.

**Pull request merged**